### PR TITLE
Allow override proxy config from agent

### PR DESCRIPTION
**Pull request by howardjohn**
The goal here is to allow overrides of proxy config per workload. This is in the context of https://docs.google.com/document/d/1JJXvQSINaAww-Hl_cdfs1QnkvfaIb6omNqzg4RXcQRY and https://github.com/istio/istio/issues/22871

This adds a new hierarchy of overrides for proxyconfig. Ultimately, the user would configure these overrides through annotations on the workload (sidecar or gateway) - the rest is implementation details. For example:
```
       annotations:
          sidecar.istio.io/proxyConfig: |-
            discoveryAddress: foo:123
```

possible future changes
* (DONE - another PR) pass all proxyConfig directly back to Pilot in node metadata

### CODE REVIEW

**Comment by howardjohn** [comment]
This is ready to review, the only thing blocking is merging the new annotation to the API repo which I will do if this is approved

File: pilot/cmd/pilot-agent/config.go [hunk 1]
```diff
@@ -59,7 +63,19 @@ func getTLSCerts(pc meshconfig.ProxyConfig) []string {
...
+	}
+	var fileMeshContents string
+	if fileExists(meshConfigFile) {
+		contents, err := ioutil.ReadFile(meshConfigFile)
+		if err != nil {
+			return meshconfig.ProxyConfig{}, fmt.Errorf("failed to read mesh config file %v: %v", meshConfigFile, err)
+		}
+		fileMeshContents = string(contents)
+	}
+	meshConfig, err := getMeshConfig(fileMeshContents, annotations[inject.MeshConfigAnnotation])
```
Comment (ramaraochavali) [inline pilot/cmd/pilot-agent/config.go:78]: Where is this `meshConfig` used? 

Also I noticed there are some missing fields in ProxyConfig compared to MeshConfig - For example file access log https://github.com/istio/api/blob/master/mesh/v1alpha1/config.proto#L151. What do we do with those fields?

Comment (howardjohn) [reply pilot/cmd/pilot-agent/config.go:78]: It's passed to the agent to configure things like access logs service, tracing, etc. This pr is to allow per proxy overrides of these

I am not sure what you mean about missing fields. ProxyConfig is a field in mesh config. Currently the proxy only uses the proxy config field, but I have chosen to pass the full mesh config in case we need to change that in the future - see linked issue for some additional thoughts

Comment (ramaraochavali) [reply pilot/cmd/pilot-agent/config.go:78]: > but I have chosen to pass the full mesh config in case we need to change that in the future 

Actually I am also looking for that - but probably missing some thing here. That is why this question. Where you passing the full mesh config to agent? Can you please point me to that?

My question specifically was
- We have `access_log_file` field in MeshConfig. Let us say user gives a value to this field in the MeshConfig
- We are getting ProxyConfig out of this `MeshConfig` in Line 84 and filling other fields and passing to Agent.

Since `access_log_file` is not available in `ProxyConfig` - how will we populate?

BTW, I agree passing full mesh config is the right thing to do for future.

Comment (howardjohn) [reply pilot/cmd/pilot-agent/config.go:78]: I understand now. Basically right now the agent never uses any fields outside of proxy config, so in line 84 we throw away all of the other stuff. So if we wanted to use the access log file somewhere, we would just need to change this to return the full thing.

We are passing the config to the agent by injection as an env var for sidecar and mounting for gateway

Comment (ramaraochavali) [reply pilot/cmd/pilot-agent/config.go:78]: > we would just need to change this to return the full thing.

Yeah. I agree. We can probably do once this lands?

Comment (howardjohn) [reply pilot/cmd/pilot-agent/config.go:78]: Its a trivial change - no need to make it until we want to use it. At some point in the code we need to ignore the mesh config and just use the proxy config so it doesn't really matter where that is right now in my opinion

Comment (ramaraochavali) [reply pilot/cmd/pilot-agent/config.go:78]: OK. I see the difference now. When I read the description specifically this `The goal here is to allow overrides of mesh config per workload` - I was assuming that we read the mesh config from file/what ever and pass it to Pilot via Node Metadata and pilot uses it apply workload specific settings. It seems that is not the case - it is to just allow/override ProxyConfig part from MeshConfig. 

Is that correct?

**Comment by howardjohn** [comment]
> I was assuming that we read the mesh config from file/what ever and pass
it to Pilot via Node Metadata and pilot uses it apply workload specific
settings. It seems that is not the case - it is to just allow/override
ProxyConfig part from MeshConfig.

Today its just to override the proxyconfig part. In the (near) future we
will likely send the whole thing as metadata to support things like the
dynamic tracing and
https://docs.google.com/document/d/1JJXvQSINaAww-Hl_cdfs1QnkvfaIb6omNqzg4RXcQRY/edit

**Comment by ramaraochavali** [comment]
Ok Thanks. Makes sense.

File: pilot/cmd/pilot-agent/config.go [hunk 2]
```diff
@@ -85,13 +101,55 @@ func constructProxyConfig() (meshconfig.ProxyConfig, error) {
...
+		}
+		mc = *annotationMesh
+	}
+
+	return mc, nil
+}
+
+func fileExists(path string) bool {
+	if _, err := os.Stat(path); os.IsNotExist(err) {
+		log.Errorf("howardjohn: %v %v", path, err)
```
Comment (nrjpoddar) [inline pilot/cmd/pilot-agent/config.go:147]: I'm guessing this was supposed to be removed?

File: pilot/cmd/pilot-agent/config.go [hunk 3]
```diff
@@ -85,13 +101,55 @@ func constructProxyConfig() (meshconfig.ProxyConfig, error) {
...
 
+// getMeshConfig gets the mesh config to use for proxy configuration
+// 1. First we take the default config
+// 2. Then we apply any settings from file (this comes from gateway mounting configmap)
+// 3. Then we apply settings from environment variable (this comes from sidecar injection sticking meshconfig here)
+// 4. Then we apply overrides from annotation (this comes from annotation on gateway, passed through downward API)
+func getMeshConfig(fileOverride, annotationOverride string) (meshconfig.MeshConfig, error) {
+	mc := mesh.DefaultMeshConfig()
+
+	if fileOverride != "" {
```
Comment (nrjpoddar) [inline pilot/cmd/pilot-agent/config.go:115]: You inverted the order from the old code; I don't think it matters as we shouldn't have file & env both set in any one workload but why did you change it?

**Comment by kyessenov** [comment]
How do you decide the merge semantics for individual fields? For extra stat tags we want addition (in the sense of `root + "," + per_proxy`).

**Comment by howardjohn** [comment]
> How do you decide the merge semantics for individual fields? For extra stat tags we want addition (in the sense of root + "," + per_proxy).

This is using the existing meshconfig merge logic, meaning replace. I don't think there is a universal right answer though, as you could just as easily say you want to override the root (or if not for this field, another field). If we need this it might make sense to have a specific tag annotation or if we don't allow `root` to be modified it can be in the code instead of as a default value for the field. What do you think?

**Comment by kyessenov** [comment]
@howardjohn I think the most important thing is to explain how merge is done. We don't have any security sensitive fields (hopefully) so it's OK to allow replacement. But it gets very confusing if you add annotations to the mix.

**Comment by nrjpoddar** [comment]
@howardjohn can we change the logic to only supply ProxyConfig instead of the entire MeshConfig during injection? Is that too big a change close to 1.6?

File: pilot/cmd/pilot-agent/config.go (hunk 3 above)
Comment (howardjohn) [reply pilot/cmd/pilot-agent/config.go:115]: The previous logic was not merging all settings, it was env OR file. Arguable we could do the same here but it seemed simpler to have everything just applying on top of the previous. As you mentioned the order doesn't matter care too much which one goes first - I think file < env < annotation just made more sense to me, I can change it if preferred

**Comment by howardjohn** [comment]
> can we change the logic to only supply ProxyConfig instead of the entire MeshConfig during injection? Is that too big a change close to 1.6?

I think we can/should do this based on recent discussions. Ill send a separate PR to do this (in parallel)

> I think the most important thing is to explain how merge is done. 

Definitely. we will need to merge the annotation in istio/api as well once this is ready so that will be a good place for that - ill also put some in the code though

File: pilot/cmd/pilot-agent/config_test.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+	"github.com/gogo/protobuf/types"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+
+	"istio.io/istio/pkg/config/mesh"
+)
+
+func TestGetMeshConfig(t *testing.T) {
+	overrides := `
+defaultConfig:
```
Comment (nrjpoddar) [inline pilot/cmd/pilot-agent/config_test.go:31]: Just for my sanity, can we add a test case for a field with arrays and ensure that these are overrides/replacements and not merged.

File: pilot/cmd/pilot-agent/config.go (hunk 3 above)
Comment (nrjpoddar) [reply pilot/cmd/pilot-agent/config.go:118]: Well, this is not a user facing API but implementation so we should be free to change it. Keep it the way you have, it's cleaner.

File: pilot/cmd/pilot-agent/config_test.go (hunk 4 above)
Comment (howardjohn) [reply pilot/cmd/pilot-agent/config_test.go:31]: I would but one does not exist right now, seems like Kuat's change hasn't made it through yet. I don't have a `map` field though

Comment (nrjpoddar) [reply pilot/cmd/pilot-agent/config_test.go:31]: Pick your favorite collection and add a test for it. If no collections exists, let's tag the PR adding it to also add the test!

Comment (howardjohn) [reply pilot/cmd/pilot-agent/config_test.go:31]: Yeah there is none existing in proxyConfig. I'll make sure we add one when that change gets in

**Comment by howardjohn** [comment]
> This change has a weird effect that if a user happened to have an annotation like this and upgraded Istio & restarted the workload after upgrade, the pod won't come up now if the annotation values is not compliant with the current schema. This will also happen to any user who has added annotations after 1.6 and then we augment ProxyConfig schema in a non-backward compatible way.

That is a good point. Currently we log the error and return it in the webhook response which I think is the best we can do here. Theres also a metric for invalid injections. Ideally we don't make backwards incompatible changes as well :slightly_smiling_face: - unknown fields are not an error so even removing stuff is fine. I am not sure its much different than any of the existing annotations, really? We also have `istioctl analyze` that will check these annotations to handle gateway case.

**Comment by nrjpoddar** [comment]
> > This change has a weird effect that if a user happened to have an annotation like this and upgraded Istio & restarted the workload after upgrade, the pod won't come up now if the annotation values is not compliant with the current schema. This will also happen to any user who has added annotations after 1.6 and then we augment ProxyConfig schema in a non-backward compatible way.
> 
> That is a good point. Currently we log the error and return it in the webhook response which I think is the best we can do here. Theres also a metric for invalid injections. Ideally we don't make backwards incompatible changes as well 🙂 - unknown fields are not an error so even removing stuff is fine. I am not sure its much different than any of the existing annotations, really? We also have `istioctl analyze` that will check these annotations to handle gateway case.

I was worried about Gateways but in a way it's better that wrong configuration will stop the workload from coming up. The feedback to the user would be a problem as they would have to inspect logs to understand what happened or run `analyze`. Is `analyze` functional for your changes?

**Comment by howardjohn** [comment]
> Is analyze functional for your changes?

Yes, its already using the annotation validation map that is utilized here as well. I think in gateway case it would fail in the startup and log some error about mesh config not being valid though now that I think about it, which is also fine to fail fast
